controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/TruncateStreamTask.java

```diff
@@ -61,7 +61,11 @@ public TruncateStreamTask(final StreamMetadataTasks streamMetadataTasks,
         return streamMetadataStore.getTruncationRecord(scope, stream, true, context, executor)
                 .thenCompose(property -> {
                     if (!property.isUpdating()) {
-                        throw new TaskExceptions.StartException("Truncate Stream not started yet.");
+                        // if the state is TRUNCATING but the truncation record is not updating, we should reset the state to ACTIVE.
+                        return streamMetadataStore.resetStateConditionally(scope, stream, State.TRUNCATING, context, executor)
```

Comment:
A couple of clarification questions:
1- This seems to be some form of rolling forward. To roll forward, we must be absolutely sure that this is the only possible state transition, otherwise we risk having an inconsistent state. Can you convince me that this is the only possible state transition at this point.

2- In the description of the PR, you say that this is already addressed for scaling operations. However, looking at `ScaleOperationTask`, I don't see the same pattern. Maybe that's ok, but I'd like to understand why.

Reply:
>  1- This seems to be some form of rolling forward. To roll forward, we must be absolutely sure that this is the only possible state transition, otherwise we risk having an inconsistent state. Can you convince me that this is the only possible state transition at this point.

for each type of processing we do things in following way:
1. set some metadata barrier specific to the type of processing
2. set state to identify the processing
3. process
4. remove metadata barrier
5. reset state to active

`1` can be done by manual call (for example manual scale and manual truncation apis) or by some background system processing.

If we ever find "state" without its corresponding metadata barrier, it means the processing was already completed.  So we can reset the state. 
For example, if the state is SCALING but there is no epochTransitionRecord, it means scaling completed and we can reset the state to ACTIVE. 
Similarly, if the state is TRUNCATING but the truncationRecord's updating flag is false, it means previous truncation completed and we can reset the state to ACTIVE. 

The method `resetStateConditionally` checks if the state is "desired state" then only reset it to ACTIVE. 

> 2- In the description of the PR, you say that this is already addressed for scaling operations. However, looking at ScaleOperationTask, I don't see the same pattern. Maybe that's ok, but I'd like to understand why.
The same thing for scale is done in PersistentStreamBase.createNewSegments where we fetch epochTransition and if it is null but the state is SCALING, we reset it to ACTIVE